src/cobra/io/mat.py

```diff
@@ -284,76 +453,123 @@ def from_mat_struct(
-    for i, name in enumerate(m["rxns"][0, 0]):
-        new_reaction = Reaction()
-        new_reaction.id = str(name[0][0]).strip()
-        new_reaction.bounds = float(m["lb"][0, 0][i][0]), float(m["ub"][0, 0][i][0])
-        if np.isinf(new_reaction.lower_bound) and new_reaction.lower_bound < 0:
-            new_reaction.lower_bound = -inf
-        if np.isinf(new_reaction.upper_bound) and new_reaction.upper_bound > 0:
-            new_reaction.upper_bound = inf
-        if c_vec is not None:
-            coefficients[new_reaction] = float(c_vec[i][0])
-        try:
-            new_reaction.gene_reaction_rule = str(m["grRules"][0, 0][i][0][0]).strip()
-        except (IndexError, ValueError):
-            # TODO: use custom cobra exception to handle exception
-            pass
-        try:
-            new_reaction.name = str(m["rxnNames"][0, 0][i][0][0]).strip()
-        except (IndexError, ValueError):
-            # TODO: use custom cobra exception to handle exception
-            pass
-        try:
-            new_reaction.subsystem = str(m["subSystems"][0, 0][i][0][0]).strip()
-        except (IndexError, ValueError):
-            # TODO: use custom cobra exception to handle exception
-            pass
+    rxn_ids = _cell_to_str_list(m["rxns"][0, 0])
+    rxn_lbs = _cell_to_float_list(m["lb"][0, 0])
+    rxn_lbs = [-inf if np.isinf(x) and x < 0 else x for x in rxn_lbs]
+    rxn_ubs = _cell_to_float_list(m["ub"][0, 0])
+    rxn_ubs = [-inf if np.isinf(x) and x > 0 else x for x in rxn_ubs]
+    rxn_gene_rules, rxn_names, rxn_subsystems = None, None, None
+    try:
+        rxn_gene_rules = _cell_to_str_list(m["grRules"][0, 0], "")
+    except (IndexError, ValueError):
+        # TODO: use custom cobra exception to handle exception
+        pass
+    try:
+        rxn_names = _cell_to_str_list(m["rxnNames"][0, 0], "")
+    except (IndexError, ValueError):
+        # TODO: use custom cobra exception to handle exception
+        pass
+    try:
+        rxn_subsystems = _cell_to_str_list(m["subSystems"][0, 0])
+    except (IndexError, ValueError):
+        # TODO: use custom cobra exception to handle exception
+        pass
+    for i in range(len(rxn_ids)):
+        new_reaction = Reaction(
+            id=rxn_ids[i], lower_bound=rxn_lbs[i], upper_bound=rxn_ubs[i]
+        )
+        if rxn_names:
+            new_reaction.name = rxn_names[i]
+        if rxn_subsystems:
+            new_reaction.subsystem = rxn_subsystems[i]
+        if rxn_gene_rules:
+            new_reaction.gene_reaction_rule = rxn_gene_rules[i]
         new_reactions.append(new_reaction)
+    mat_annotations(new_reactions, m, _d_replace=D_REACTION)
     model.add_reactions(new_reactions)
-    set_objective(model, coefficients)
 
-    coo = scipy_sparse.coo_matrix(m["S"][0, 0])
-    for i, j, v in zip(coo.row, coo.col, coo.data):
-        model.reactions[j].add_metabolites({model.metabolites[i]: v})
+    csc = scipy_sparse.csc_matrix(m["S"][0, 0])
```

Comment:
We should try out what happens if we do this before running `model.add_reactions(new_reactions). There are some optimizations that kick in if you add reactions with assigned stoichiometry in bulk.

Reply:
That is (with the other suggestions you made) much much faster